src/index.js

```diff
@@ -1,83 +1,137 @@
 'use strict'
 
 const net = require('net')
-const toPull = require('stream-to-pull-stream')
 const mafmt = require('mafmt')
 const withIs = require('class-is')
-const includes = require('lodash.includes')
-const isFunction = require('lodash.isfunction')
-const Connection = require('interface-connection').Connection
-const once = require('once')
-const debug = require('debug')
-const log = debug('libp2p:tcp:dial')
-
+const errCode = require('err-code')
+const log = require('debug')('libp2p:tcp')
+const toConnection = require('./socket-to-conn')
 const createListener = require('./listener')
+const { AbortError } = require('abortable-iterator')
+const { CODE_CIRCUIT, CODE_P2P } = require('./constants')
+const assert = require('assert')
 
-function noop () {}
-
+/**
+ * @class TCP
+ */
 class TCP {
-  dial (ma, options, callback) {
-    if (isFunction(options)) {
-      callback = options
-      options = {}
-    }
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {Upgrader} options.upgrader
+   */
+  constructor ({ upgrader }) {
+    assert(upgrader, 'An upgrader must be provided. See https://github.com/libp2p/interface-transport#upgrader.')
```

Comment:
The readme example has this class instantiable without any options.

Reply:
Doing a PR shortly for another thing, I will include that there.